### PR TITLE
fix data races

### PR DESCRIPTION
**Pull request by syntrust**

### CODE REVIEW

File: core/rootblock_validator.go [hunk 1]
```diff
@@ -123,7 +124,10 @@ func (v *RootBlockValidator) ValidateBlock(block types.IBlock, force bool) error
 		parentHeader = mheader
 		shardIdToMinorHeadersMap[fullShardId] = append(shardIdToMinorHeadersMap[fullShardId], mheader)
 	}
-
+	for k, v := range shardIdToMinorHeadersMap {
```
Comment (pingke) [inline core/rootblock_validator.go]: remove?

File: p2p/discover/table.go [hunk 2]
```diff
@@ -475,13 +477,13 @@ func (tab *Table) doRevalidate(done chan<- struct{}) {
...
 
+	tab.mutex.Lock()
 	inBlackList := false
 	if tab.checkDialBlackList(last.IP().String()) {
 		log.Warn("black node", "b", bi, "id", last.ID(), "address", last.addr().String())
 		inBlackList = true
 	}
 
-	tab.mutex.Lock()
 	defer tab.mutex.Unlock()
```
Comment (pingke) [inline p2p/discover/table.go]: move to link 481

File: cmd/stats/stats.go [hunk 3]
```diff
@@ -50,21 +50,24 @@ func basic(clt jsonrpc.RPCClient, ip string) string {
 	return msg
 }
 
-func queryStats(client jsonrpc.RPCClient, interval *uint) {
-	titles := []string{"Timestamp\t", "Syncing", "TPS", "Pend.TX", "Conf.TX", "BPS", "SBPS", "CPU", "ROOT", "CHAIN/SHARD-HEIGHT"}
+func queryStats(client jsonrpc.RPCClient, interval *uint, shards bool) {
```
[the remark this answers is not in the document]
Comment (syntrust) [reply cmd/stats/stats.go]: show shards height or not

File: common/utils.go [hunk 4]
```diff
@@ -109,8 +109,9 @@ func (c *ConstMinorBlockRewardCalculator) GetBlockReward() *big.Int {
 }
 
 func BigIntMulBigRat(bigInt *big.Int, bigRat *big.Rat) *big.Int {
-	ans := new(big.Int).Mul(bigInt, bigRat.Num())
-	ans.Div(ans, bigRat.Denom())
+	bigRat1 := new(big.Rat).Set(bigRat)
```
[the remark this answers is not in the document]
Comment (syntrust) [reply common/utils.go]: copy so Denom() cannot modify the origin

File: cluster/service/service.go [hunk 5]
```diff
@@ -20,18 +20,23 @@ type ServiceContext struct {
...
+func (ctx *ServiceContext) SetTimestamp(timestamp time.Time) {
 	ctx.mu.Lock()
-}
-func (ctx *ServiceContext) UnlockTimestamp() {
+	ctx.timestamp = timestamp
 	ctx.mu.Unlock()
 }
 
+func (ctx *ServiceContext) GetTimestamp() time.Time {
+	ctx.mu.Lock()
```
Comment (pingke) [inline cluster/service/service.go]: RLock?

File: core/minorblockchain.go [hunk 6]
```diff
@@ -1051,11 +1052,13 @@ func (m *MinorBlockChain) InsertChainForDeposits(chain []types.IBlock, isCheckDB
 	m.wg.Done()
 
 	m.PostChainEvents(events, logs)
+	m.mu.Lock()
```
[the remark this answers is not in the document]
Comment (syntrust) [reply core/minorblockchain.go]: need lock for set m.confirmedHeaderTip

File: cluster/service/service.go (hunk 5 above)
Comment (syntrust) [reply cluster/service/service.go]: done

**Pull request merged**